python/lsst/daf/butler/core/dimensions/standard.py

```diff
@@ -0,0 +1,411 @@
+        are fetched from the database.
+    viewOf : `str`, optional
+        Name of another `StandardDimension` or `StandardDimensionCombination`
+        whose records this element's records summarize.
+    uniqueKeys : `NamedValueAbstractSet [ `ddl.FieldSpec` ]
+        Fields that can each be used to uniquely identify this dimension (given
+        values for all required dimensions).  The first of these is used as
+        (part of) this dimension's table's primary key, which others used to
+        define unique constraints.  Should be empty for
+        `StandardDimensionCombination` definitions.
+    """
+    def __init__(
+        self,
+        name: str,
+        required: Set[str],
+        implied: Set[str],
+        metadata: Iterable[ddl.FieldSpec] = (),
+        cached: bool = False,
+        viewOf: Optional[str] = None,
+        uniqueKeys: Iterable[ddl.FieldSpec] = (),
+        alwaysJoin: bool = False,
+    ):
+        super().__init__(name)
+        self._required = required
+        self._implied = implied
+        self._metadata = NamedValueSet(metadata).freeze()
+        self._cached = cached
+        self._viewOf = viewOf
+        self._uniqueKeys = NamedValueSet(uniqueKeys).freeze()
+        self._alwaysJoin = alwaysJoin
+
+    def hasDependenciesIn(self, others: AbstractSet[str]) -> bool:
+        # Docstring inherited from DimensionConstructionVisitor.
+        return not (
+            self._required.isdisjoint(others)
+            and self._implied.isdisjoint(others)
+        )
+
+    def visit(self, builder: DimensionConstructionBuilder) -> None:
+        # Docstring inherited from DimensionConstructionVisitor.
+        # Expand required dependencies.
+        for name in tuple(self._required):  # iterate over copy
+            self._required.update(builder.dimensions[name].required.names)
+        # Transform required and implied Dimension names into instances,
+        # and reorder to match builder's order.
+        required: NamedValueSet[Dimension] = NamedValueSet()
+        implied: NamedValueSet[Dimension] = NamedValueSet()
+        for dimension in builder.dimensions:
+            if dimension.name in self._required:
+                required.add(dimension)
+            if dimension.name in self._implied:
+                implied.add(dimension)
+
+        if self._uniqueKeys:
+            if self._alwaysJoin:
+                raise RuntimeError(f"'alwaysJoin' is not a valid option for Dimension object {self.name}.")
+            # Special handling for creating Dimension instances.
+            dimension = StandardDimension(
+                self.name,
+                required=required,
+                implied=implied.freeze(),
+                metadata=self._metadata,
+                cached=self._cached,
+                viewOf=self._viewOf,
+                uniqueKeys=self._uniqueKeys,
+            )
+            required.add(dimension)
+            required.freeze()
```

Comment:
This is potentially problematic if `StandardDimension` decides to make a copy of `required`. Can this be done in `StandardDimension` constructor?

Reply:
Yes, don't know why I didn't think of that solution here.